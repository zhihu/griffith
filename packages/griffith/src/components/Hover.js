import React from 'react'
import {sequence} from 'griffith-utils'
import noop from 'lodash/noop'

export default class Hover extends React.Component {
  state = {
    isHovered: false,
  }

  handlePointerEnter = () => {
    this.setState({isHovered: true})
  }

  handlePointerLeave = () => {
    this.setState({isHovered: false})
  }

  render() {
    const {onMouseEnter, onMouseLeave, children, ...rest} = this.props
    const {isHovered} = this.state
    return (
      <div
        {...rest}
        onMouseEnter={sequence(this.handlePointerEnter, onMouseEnter || noop)}
        onMouseLeave={sequence(this.handlePointerLeave, onMouseLeave || noop)}
      >
        {children(isHovered)}
      </div>
    )
  }
}
