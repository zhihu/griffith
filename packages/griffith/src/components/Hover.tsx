import React from 'react'
import useBoolean from '../hooks/useBoolean'

type Props = React.HTMLAttributes<HTMLDivElement> & {
  children: (isHover: boolean) => React.ReactElement
}

const Hover: React.FC<Props> = ({
  children,
  onMouseEnter,
  onMouseLeave,
  ...rest
}) => {
  const [isHovered, isHoveredSwitch] = useBoolean()

  return (
    <div
      {...rest}
      onMouseEnter={(e) => {
        isHoveredSwitch.on()
        onMouseEnter?.(e)
      }}
      onMouseLeave={(e) => {
        isHoveredSwitch.off()
        onMouseLeave?.(e)
      }}
    >
      {children(isHovered)}
    </div>
  )
}

export default Hover
