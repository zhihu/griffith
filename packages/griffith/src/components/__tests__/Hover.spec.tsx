import React from 'react'
import {render, fireEvent, screen} from '@testing-library/react'
import Hover from '../Hover'

describe('Hover', () => {
  it('get Hover component', () => {
    const handleMouseEnter = jest.fn()
    const handleMouseLeave = jest.fn()
    let hovering = false
    const result = render(
      <Hover
        className="hover"
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
      >
        {(isHovered) => {
          hovering = isHovered
          return <button>{isHovered ? 'button' : 'input'}</button>
        }}
      </Hover>
    )
    expect(result.container).toMatchSnapshot()
    expect(hovering).toBeFalsy()

    // PointerEnter
    fireEvent.mouseEnter(screen.getByRole('button'))
    expect(handleMouseEnter).toBeCalled()
    expect(hovering).toBeTruthy()

    // PointerLeave
    fireEvent.mouseLeave(screen.getByRole('button'))
    expect(handleMouseLeave).toBeCalled()
    expect(hovering).toBeFalsy()
  })
})
