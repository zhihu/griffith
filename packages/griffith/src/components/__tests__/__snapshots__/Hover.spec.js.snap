// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hover get Hover component 1`] = `
<div
  className="hover"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button>
    input
  </button>
</div>
`;
